### PR TITLE
[stable/traefik] Optionally install a DaemonSet instead of a Deployment

### PR DESCRIPTION
**Pull request by ravishivt**
This removes the requirement that the cluster be installed to a cloud provider that supports `LoadBalancer`.  For clusters that don't support LoadBalancers, a DaemonSet is a better option since all nodes can be used to reach services instead of only the node that currently has the `traefik` pod scheduled.

The DaemonSet uses `hostNetwork: true` due to https://github.com/kubernetes/kubernetes/issues/23920.  Note that traefik's kubernetes installation example [also uses hostNetwork](https://github.com/containous/traefik/blob/master/examples/k8s/traefik.yaml#L21).

Also added is option to configure traefik's dashboard port since the default 8080 will conflict with the default `kube-apiserver` port.

### CODE REVIEW

**Comment by ravishivt** [comment]
cc @krancour 

I'm new to Go templating so any suggestions on cleaning that up are welcome.  I couldn't seem to find a ternary operation.

**Comment by krancour** [comment]
Thanks for the PR. Will look at this more closely on Monday. Something to watch out for, however, as has been noted recently in other Traefik chart PRs... this won't scale Traefik _properly_, as it's intended to use a distributed k/v store to do leader election. This will "work," but mileage will vary quite a bit.

**Comment by krancour** [comment]
Thanks for the PR. Will look at this more closely on Monday. Something to watch out for, however, as has been noted recently in other Traefik chart PRs... this won't scale Traefik _properly_, as it's intended to use a distributed k/v store to do leader election. This will "work," but mileage will vary quite a bit.

**Comment by cknowles** [comment]
I suppose it's also worth noting the currently limitations of DaemonSet such as [rolling updates](https://github.com/kubernetes/kubernetes/issues/22543) not being supported yet.

**Comment by ravishivt** [comment]
Alternatively, we could implement a proxy daemonset, similar to how kube-registry-proxy proxies to the kube-registry service.  That way, we could still have the simple one traefik pod instance.  In this case, it would be a proxy to a proxy.  😄 

The problem with this approach is that the daemonset would still need to be `hostNetwork: true` and pods with hostNetwork can't resolve cluster domain names due to https://github.com/kubernetes/kubernetes/issues/17406 so we'd have to hardcode the traefik service IP, which isn't going to work long run.

**Comment by krancour** [comment]
> This removes the requirement that the cluster be installed to a cloud provider that supports LoadBalancer. For clusters that don't support LoadBalancers, a DaemonSet is a better option since all nodes can be used to reach services instead of only the node that currently has the traefik pod scheduled.

Got to thinking about this more and I believe the above is already achievable without falling back to this unconventional option.

> Alternatively, we could implement a proxy daemonset, similar to how kube-registry-proxy proxies to the kube-registry service.

k8s can already do that for you. (i.e. kube-proxy on each node proxies your service.) From the documentation:

https://kubernetes.io/docs/user-guide/services/#type-nodeport

> If you set the type field to "NodePort", the Kubernetes master will allocate a port from a flag-configured range (default: 30000-32767), and each Node will proxy that port (the same port number on every Node) into your Service. That port will be reported in your Service’s spec.ports[*].nodePort field.

(fwiw, `Type: LoadBalancer` _also_ does that, in addition to all the stuff the cloud provider will do.)

This chart _already_ lets you select an alternative service type if `LoadBalancer` doesn't suit you.

See here:

https://github.com/kubernetes/charts/blob/feccd08c86532ec301ab5ea623173184dad04ce6/stable/traefik/values.yaml#L3

And:

https://github.com/kubernetes/charts/blob/feccd08c86532ec301ab5ea623173184dad04ce6/stable/traefik/README.md#configuration

Bottom line, imho, daemonsets aren't a good way to achieve this. Could we consider closing this?

**Comment by ravishivt** [comment]
Thanks for the comments but I disagree on the viability of `nodePort`.  Although I agree it would theoretically work, the purpose of traefik is to bind on common 80/443 ports and then forward those requests to registered services on more abstract IPs and ports.  It kind of defeats the purpose if I have to tell someone to reach my service at jenkins.example.com:30165 instead of just jenkins.example.com.

**Comment by krancour** [comment]
> the purpose of traefik is to bind on common 80/443 ports and then forward those requests to registered services on more abstract IPs and ports

I don't think it's clear that that's Traefik's primary purpose. I doubt its own authors would say so.

>  kind of defeats the purpose if I have to tell someone to reach my service at jenkins.example.com:30165 instead of just jenkins.example.com.

That's not at all what you do. This thread started with an assumption that someone using this proposed feature would be someone running somewhere where the cloud provider no-ops the creation of a load balancer for services of `Type: LoadBalancer`. So, let's say, bare metal, for instance. In such cases, aren't you running your _own_ load balancer? HAProxy, perhaps? Instead of configuring that to pass traffic to your nodes on ports 80 and 443, you configure it to pass traffic to your nodes on the port(s) for which the kube-proxy on each node would be forwarding traffic to Traefik... end-users never, ever know that the traffic hits each node on some other port.

**Comment by ravishivt** [comment]
> I don't think it's clear that that's Traefik's primary purpose. I doubt its own authors would say so.

You're right.  I should have said "my purpose for using traefik", although I do believe it is a common use case to want traefik to proxy from standard ports to non-standard ones.  Common enough that `traefik` uses `hostNetwork: true` in their own [kubernetes setup examples](https://github.com/containous/traefik/blob/master/examples/k8s/traefik.yaml#L21).

> So, let's say, bare metal, for instance. In such cases, aren't you running your own load balancer? HAProxy, perhaps? Instead of configuring that to pass traffic to your nodes on ports 80 and 443, you configure it to pass traffic to your nodes on the port(s) for which the kube-proxy on each node would be forwarding traffic to Traefik... end-users never, ever know that the traffic hits each node on some other port.

For a user who just wants to kick the tires, they have the option to point DNS to one of the nodes without having to worry about which node the pod is currently scheduled on.  Slightly better than that, they could set up DNS round-robin for a quick and dirty load balancing option.

I agree that the DaemonSet approach has its drawbacks but it lowers the barrier to entry to get traefik "working".  Also, none of this would be necessary if/when kubernetes implements `hostPort` support for CNI enabled clusters.

**Comment by krancour** [comment]
Ok. Your point about DNS round robining being an easy way to kick the tires on this is well taken.

That being said, there's something about this that still doesn't sit well with me. It feels very heavy handed to say that based on a cluster's inability to provision a load balancer, the solution is to run Traefik on every node. It's not helped by the fact that, as I have mentioned, this isn't the _right_ way to scale Traefik. Scaling it properly requires it to have access to a distributed key/value store to help facilitate leader election. What I'm getting around to here is that for the sake of allowing on bare metal to kick the tires more easily, you want to enable something that really shouldn't be done.

There's actually a separate PR that implenents the naive approach to scaling-- i.e. just makes number of replicas configurable. You will see I had some similar pushback on that PR for the same reason. What is different between that PR and this is that the accompanying documentation is clear that it is a naive approach to scaling and that your mileage may vary. In this PR, the doc changes seem to mention daemonsets as a viable option, and makes no mention of any caveats.

As for this _and_ the scaling PR, I have to start asking myself, as the number of knobs and dials an operator can turn on this chart increases, what is the value of each one? Is it widely useful? Or is it useful, perhaps, only to a very few. Does the usefulness come at a price? i.e. by encouraging people to do something they really shouldn't. I think it's a logical consideration to make that perhaps there is a certain category of features people may want in rare cases and that those operators might be better off consuming those from their own fork of this chart.

I'd like to get another maintainer's perspective on this. @lachie83, could you weigh in, please?

**Comment by krancour** [comment]
https://github.com/kubernetes/charts/pull/431 is the other PR I was referencing.

**Comment by ravishivt** [comment]
I can update the documentation to better inform users of the drawbacks of enabling the daemonset option.  I will wait on if we decide this PR is wanted at all though.

Another alternative that would eliminate the need for a DaemonSet altogether is to add a `nodeSelector` option to the current Deployment.  That way, users get a predictable node IP that they can add to DNS.  The disadvantage is that there is no redundancy.

**Comment by ravishivt** [comment]
Interestingly, this [article](https://hackernoon.com/kubernetes-ingress-controllers-and-traefik-a32648a4ae95#.3ct30418p) advocates the use of a Traefik `DaemonSet` in addition with `nodeSelector`.  It doesn't explain why you would want do that very well though.  I'm guessing you start with one node and then can add a second if redundancy is more important than the multiple instance drawbacks (multi-cert requests, daemonset rolling updates, etc).  I'm not vouching for this approach, it's just another data point that highlights it's a problem other people are facing / trying to tackle.

**Comment by viglesiasce** [comment]
Ill chime in on this one 😄 

I agree that it is heavy handed but can also sympathize with the bare metal use case.

I believe the right approach would be to set the traefik service as NodePort with a customizable host port (if set). If your cluster allows the service to bind to low ports then you are all set for your use case @ravishivt. 

Given that [traefik documentation for Clustering/HA](https://docs.traefik.io/user-guide/cluster/) has a KV store as a prerequisite I'd be led to believe steering people towards a clustered deployment that doesn't include that is a bad idea.

**Comment by krancour** [comment]
> I believe the right approach would be to set the traefik service as NodePort with a customizable host port (if set). If your cluster allows the service to bind to low ports then you are all set for your use case @ravishivt.

@ravishivt would you consider submitting a PR for that instead?

**Comment by ravishivt** [comment]
@viglesiasce The problem is that using HostPort [doesn't currently work](https://github.com/kubernetes/kubernetes/issues/23920) on clusters with CNI enabled.  So it won't help many bare-metal users.

**Comment by ravishivt** [comment]
If we are decided that naive multi-traefik instances are a bad practice that we don't want to advocate using, maybe we hold off on this PR until one of the following happens:

1. https://github.com/kubernetes/kubernetes/issues/23920 is fixed so we can use hostPort
2. Some kind soul updates this chart to properly support multi-traefik instances with KV stores.  Also hopefully by that time, rolling-update support is added to daemonset.
3. https://github.com/kubernetes/kubernetes/issues/17406 is fixed so we can add a daemonset that only does proxying to traefik's service. 

For temporary relief, what about my alternative proposed [a few comments ago](https://github.com/kubernetes/charts/pull/477#issuecomment-275287155), where we add a nodeSelector variable plus set `hostNetwork: true`?

**Comment by lachie83** [comment]
@krancour @ravishivt @viglesiasce What's the latest on this? Given that this is a configurable change that doesn't change default behavior I'm okay with it. I know several folks running ingress controllers as daemonsets especially in bare metal clusters. The only part that gets you here is the inability to upgrade daemonsets.

**Comment by krancour** [comment]
I don't personally feel good about features that don't work well with other features. e.g. This won't work well with ACME. But we've set a precedent in #431 that that is ok since ACME isn't a feature that's enabled by default. In light of that, I'll shelve my objections and re-review this on Monday so we can get it merged.

**Comment by ravishivt** [comment]
I updated the PR to add disclaimers about multi-traefik instances.  I copied the disclaimers from #431 and modified them to cover both replicas and daemonset.  This is under the assumption that both will eventually be merged.  I can change it to specifically disclaim only about daemonset if preferred.

**Comment by lachie83** [comment]
@ravishivt @krancour Are we happy for this to be merged?

**Comment by krancour** [comment]
@lachie83 I'll re-review this on Monday, but at minimum, it appears this is in need of a rebase.

**Comment by lachie83** [comment]
@krancour rebased. Have at it

**Comment by seanknox** [comment]
@ravishivt can you bump the minor version of the chart?

File: stable/traefik/README.md [hunk 1]
```diff
@@ -18,8 +18,6 @@ resources _cluster-wide_.
 ## Prerequisites
 
 - Kubernetes 1.4+ with Beta APIs enabled
-- You are deploying the chart to a cluster with a cloud provider capable of provisioning an
-external load balancer (e.g. AWS or GKE)
```
Comment (krancour) [inline stable/traefik/README.md:22]: I know the idea of managing this with a daemonset technically eliminates this prerequisite, but since the whole daemonset thing is sort of a "use at your own risk" option that I really still don't endorse, I would rather this prereq weren't struck out. I would suggest putting it back in _or_ amending it to say this prereq can be ignored if and only if you're willing to adopt the naive approach of scaling Traefik via daemonset.

File: stable/traefik/templates/deployment.yaml [hunk 2]
```diff
@@ -62,7 +71,7 @@ spec:
         - containerPort: 80
         - containerPort: 443
         {{- if .Values.dashboard.enabled }}
-        - containerPort: 8080
+        - containerPort: {{ .Values.dashboard.port }}
```
Comment (krancour) [inline stable/traefik/templates/deployment.yaml:74]: What is the reason for this particular change? This looks unrelated to the "run using a daemonset" feature. And although this may even be a common thing to allow users to configure, I'm unaware of compelling reasons for anyone to concern themselves with what specific port a process listens on _within_ a container. I would remove this change, and if you can make a compelling case for it, please make it a separate PR.

File: stable/traefik/values.yaml [hunk 3]
```diff
@@ -27,6 +27,7 @@ acme:
     size: 1Gi
 dashboard:
   enabled: false
+  port: 8080
```
Comment (krancour) [inline stable/traefik/values.yaml:30]: See https://github.com/kubernetes/charts/pull/477/files#r106914606

File: stable/traefik/values.yaml [hunk 4]
```diff
@@ -41,6 +42,8 @@ service:
   #   key: value
 gzip:
   enabled: true
+daemonset:
+  enabled: false
```
Comment (krancour) [inline stable/traefik/values.yaml:46]: I think I'd prefer that we exposed an option `deploymentModel` with allowable values `deployment` and `daemonset`. I say this because, looking toward the future and a hope of supporting a more sophisticated scaling option (master/slave), I wouldn't want this choice to be binary.

File: stable/traefik/README.md [hunk 5]
```diff
@@ -86,8 +84,10 @@ The following tables lists the configurable parameters of the Traefik chart and
 | `acme.persistence.storageClass` | Type of `StorageClass` to request-- will be cluster-specific         | `nil` (uses alpha storage class annotation) |
 | `acme.persistence.accessMode`   | `ReadWriteOnce` or `ReadOnly`                                        | `ReadWriteOnce`                           |
 | `acme.persistence.size`         | Minimum size of the volume requested                                 | `1Gi`                                     |
+| `daemonset.enabled`             | Deploy DaemonSet instead of Deployment.  __NOTE:__ Full Traefik clustering with leader election is not yet supported, which can affect any configured Let's Encrypt setup; see Clustering section | `false` |
 | `dashboard.enabled`             | Whether to enable the Traefik dashboard                              | `false`                                   |
 | `dashboard.domain`              | Domain for the Traefik dashboard                                     | `traefik.example.com`                     |
+| `dashboard.port`                | Container port for the Traefik dashboard                             | `8080`                                    |
```
Comment (krancour) [inline stable/traefik/README.md:91]: See https://github.com/kubernetes/charts/pull/477/files#r106914606

File: stable/traefik/README.md [hunk 6]
```diff
@@ -86,8 +84,10 @@ The following tables lists the configurable parameters of the Traefik chart and
 | `acme.persistence.storageClass` | Type of `StorageClass` to request-- will be cluster-specific         | `nil` (uses alpha storage class annotation) |
 | `acme.persistence.accessMode`   | `ReadWriteOnce` or `ReadOnly`                                        | `ReadWriteOnce`                           |
 | `acme.persistence.size`         | Minimum size of the volume requested                                 | `1Gi`                                     |
+| `daemonset.enabled`             | Deploy DaemonSet instead of Deployment.  __NOTE:__ Full Traefik clustering with leader election is not yet supported, which can affect any configured Let's Encrypt setup; see Clustering section | `false` |
```
Comment (krancour) [inline stable/traefik/README.md:88]: See https://github.com/kubernetes/charts/pull/477/files#r106915573

File: stable/traefik/README.md [hunk 7]
```diff
@@ -114,11 +114,26 @@ $ helm install --name my-release --namespace kube-system --values values.yaml st
 
 ### Clustering / High Availability
 
-Currently it is possible to specify the number of `replicas` but the implementation is naive.
+Currently it is possible to configure this chart to run multiple Traefik instances using
+either the `replicas` option or the `daemonset.enabled` option.  Enabling either results
+in a naive Traefik implementation.
```
Comment (krancour) [inline stable/traefik/README.md:120]: This sounds too scary... I would qualify that the part that is naive is, specifically, the approach to scaling.

File: stable/traefik/README.md [hunk 8]
```diff
@@ -114,11 +114,26 @@ $ helm install --name my-release --namespace kube-system --values values.yaml st
...
+Traefik instances will work for many cases, since no leader is elected it has the consequence 
+that each node will end up requesting Let's Encrypt certificates if this is also configured.
+This will quickly cut into the very modest rate limit that Let's Encrypt enforces. 
+
+If you are installing the chart to a cluster that is not capable of provisioning an
+external load balancer (i.e. not using AWS or GKE), using `daemonset.enabled=true`
+can help make Traefik accessible outside of the cluster on common 80/443 ports.
+Configuring a DaemonSet will schedule a Traefik pod on each node with `hostNetwork` enabled 
+due to https://github.com/kubernetes/kubernetes/issues/23920.  You will likely also want
+to change the `dashboard.port` so it doesn't conflict with your master node's
```
Comment (krancour) [inline stable/traefik/README.md:135]: So... I have already registered some concern over `dashboard.port`. I see, now, after reading this bit of doc change that what you've implemented isn't in line with what you've documented here. What you've implemented gives control over the port that the dashboard listens on within the container... that's not the knob you want to expose. You want to control the port the _service_ listens on when a daemonset is used.

File: stable/traefik/README.md (hunk 1 above)
Comment (viglesiasce) [reply stable/traefik/README.md:22]: I agree with this and added it back.

**Comment by viglesiasce** [comment]
ping @ravishivt for fixes related to @krancour's feedback

**Comment by prydonius** [comment]
@ravishivt marking as stale, please update within the next week

**Comment by viglesiasce** [comment]
Closing as this has been stale for 17 days:

https://github.com/kubernetes/charts#stale-pull-requests

**Comment by kachkaev** [comment]
Ping @ravishivt. I guess what you proposing may be helpful for the bare-metal clusters indeed. It may also be used in the cloud-based clusters such as ones on AWS if the setups are small enough to avoid extra charges for a load balancer. This is currently possible with `nodePort` too, however to achieve this, the cluster has to start with a custom `service-node-port-range` (so that it includes 80 and 443).

I recently asked a [related question on Stack Overflow](https://stackoverflow.com/questions/45279433/make-k8s-services-available-via-ingress-on-an-aws-cluster-created-with-kops), but it's still unclear what's best to do in order to get a public-facing cluster with traefik without config patching after `helm install`. Would be great if you shared your thoughts!

---
UPD: RollingUpdates for DaemonSets seem to be also supported since k8s 1.7, so at least one argument against DeamonSet are no longer accurate.
